feat(dart/transform): Use the render Compiler and the DirectiveParser

kegluneq opened the pull request:
Update the `TemplateCompile` step to use the full render `Compiler`.

Provide `DirectiveMetadata` for `ViewDefinition` objects and use it to
run the `DirectiveParser` step of the render compile pipeline.

Commit:
feat(dart/transform): Use the render Compiler and the DirectiveParser

Update the `TemplateCompile` step to use the full render `Compiler`.

Provide `DirectiveMetadata` for `ViewDefinition` objects and use it to
run the `DirectiveParser` step of the render compile pipeline.

kegluneq commented:
Sorry for the large PR - let me know if you'd like me to split it up.

yjbanov commented on modules/angular2/src/dom/html_adapter.dart:
```diff
@@ -83,7 +83,14 @@ class Html5LibDomAdapter implements DomAdapter {
+      case Node.TEXT_NODE:
+        return '#text';
+      default:
+        throw 'not implemented';
```
Now that it is implemented, let's throw a message that contains the node type.

Commit:
fix(dart/transform): Handle `hostAttributes` in DirectiveMetadata

Handle `hostAttributes` in the transformer.
`hostAttributes` was introduced in 51839ca6772ed5e033c71118ce1dc78cc990246d

yjbanov commented on modules/angular2/src/transform/template_compiler/view_definition_creator.dart:
````diff
@@ -0,0 +1,232 @@
+  /// ```
+  ///
+  /// This method will look for `component.ng_meta.json`to contain the
+  /// serialized [DirectiveMetadata] `MyComponent` and any other `Directive`s
````
[DirectiveMetadata] *for* `MyComponent`

yjbanov commented on modules/angular2/src/transform/template_compiler/view_definition_creator.dart:
```diff
@@ -0,0 +1,232 @@
+}
+
+/// Visitor responsible for processing the `annotations` property of a
+/// {@link RegisterType} object and pulling out [ViewDefinition] information.
```
Does `{@link}` work in Dart?

yjbanov commented on modules/angular2/src/transform/template_compiler/view_definition_creator.dart:
```diff
@@ -0,0 +1,232 @@
+  /// These correspond to the annotations themselves.
+  @override
+  Object visitInstanceCreationExpression(InstanceCreationExpression node) {
+    if ('${node.constructorName.type}' == 'View') {
```
In the future we might want to be more precise about identifying `@View` annotations than doing a string comparison. Let's create a utility methods that use the current naive method:

```dart
bool isViewAnnotation(InstanceCreationExpression node) =>
    '${node.constructorName.type}' == 'View';
```

Later we could improve these methods to be more precise.

yjbanov commented on modules/angular2/src/transform/template_compiler/view_definition_creator.dart:
```diff
@@ -0,0 +1,232 @@
+
+  /// These correspond to the annotation parameters.
+  @override
+  Object visitNamedExpression(NamedExpression node) {
```
Does this not need a `viewDef != null` check? I imagine you can enter here while not in the context of a `@View` annotation.

kegluneq replied:
Good catch, thanks

yjbanov commented on modules/angular2/src/transform/template_compiler/view_definition_creator.dart:
```diff
@@ -0,0 +1,232 @@
+      if (node is! SimpleIdentifier && node is! PrefixedIdentifier) {
+        logger.error(
+            'Angular 2 currently only supports simple and prefixed identifiers '
+            'as values for "directives". Source: $node');
```
We also support nested lists, although it's only useful if you are referring to an external constant. Not sure if we want to do anything about it right now.

kegluneq replied:
Yeah, I've been thinking about this (see https://github.com/angular/angular/issues/1747) but I'm not sure how to support nested lists quite yet.

yjbanov commented on modules/angular2/test/transform/integration/simple_annotation_files/expected/bar.ng_meta.json:
```diff
@@ -0,0 +1 @@
+{"MyComponent":{"id":"MyComponent","selector":"[soup]","compileChildren":true,"hostListeners":{},"hostProperties":{},"properties":{},"readAttributes":[],"type":1,"version":1}}
```
Can this file be prettified?

Commit:
PR comment fixes -- SQUASH ME!
